cpp/src/fil/fil.cu

```diff
@@ -583,59 +619,56 @@ void check_params(const forest_params_t* params, bool dense)
-                                         const std::vector<float>& vector_leaf,
-                                         const int* trees,
-                                         const sparse_node16<float>* nodes,
-                                         const forest_params_t* params);
-
-template void init<sparse_node8>(const raft::handle_t& h,
-                                 forest_t* pf,
-                                 const categorical_sets& cat_sets,
-                                 const std::vector<float>& vector_leaf,
-                                 const int* trees,
-                                 const sparse_node8* nodes,
-                                 const forest_params_t* params);
-
-template void init<dense_node<float>>(const raft::handle_t& h,
-                                      forest_t* pf,
-                                      const categorical_sets& cat_sets,
-                                      const std::vector<float>& vector_leaf,
-                                      const int* trees,
-                                      const dense_node<float>* nodes,
-                                      const forest_params_t* params);
+// explicit instantiations for init()
+template void init<dense_node<float>, float>(const raft::handle_t& h,
+                                             forest_t* pf,
+                                             const categorical_sets& cat_sets,
+                                             const std::vector<float>& vector_leaf,
+                                             const int* trees,
+                                             const dense_node<float>* nodes,
+                                             const forest_params_t* params);
+template void init<sparse_node16<float>, float>(const raft::handle_t& h,
+                                                forest_t* pf,
+                                                const categorical_sets& cat_sets,
+                                                const std::vector<float>& vector_leaf,
+                                                const int* trees,
+                                                const sparse_node16<float>* nodes,
+                                                const forest_params_t* params);
+template void init<sparse_node8, float>(const raft::handle_t& h,
+                                        forest_t* pf,
+                                        const categorical_sets& cat_sets,
+                                        const std::vector<float>& vector_leaf,
+                                        const int* trees,
+                                        const sparse_node8* nodes,
+                                        const forest_params_t* params);
 
 void free(const raft::handle_t& h, forest_t f)
 {
   f->free(h);
   delete f;
 }
 
+/// part of C API - preds and data are either both pointers to float or to double
```

Comment:
I would recommend having a separate function for the C API that includes the necessary type dispatch and maintain type safety for C++ callers.

Reply:
Done (removed the comment).